clang/lib/CIR/CodeGen/CIRGenDecl.cpp

```diff
@@ -485,7 +486,8 @@ CIRGenModule::getOrCreateStaticVarDecl(const VarDecl &D,
   cir::GlobalOp GV = builder.createVersionedGlobal(
       getModule(), getLoc(D.getLocation()), Name, LTy, false, Linkage, AS);
   // TODO(cir): infer visibility from linkage in global op builder.
-  GV.setVisibility(getMLIRVisibilityFromCIRLinkage(Linkage));
+  GV.setVisibility(
+      cir::deduceMLIRVisibility(Linkage, cir::VisibilityKind::Default));
```

Comment:
My suggestion is a bit more simple and somewhat different: GV.setLinkage and Fn.setLinkage should call GV.setVisibility under the hood, no need to involve visibility kind. If you make linkage into an interface, than you don't need to implement in both places.